server.js

```diff
@@ -730,167 +729,131 @@ app.post('/submit-match',  asyncMiddleware( async (req, res, next) => {
-                                    new_best_network_flag = true;
-
-                                    fs.copyFileSync(__dirname + '/network/' + req.body.winnerhash + '.gz', __dirname + '/network/best-network.gz');
-                                    console.log("New best network copied from (fast check): " + __dirname + '/network/' + req.body.winnerhash + '.gz');
-                                }
-                            }
-                        } else {
-                            // network1 was the winner but it was no longer at the end of the pending_match queue.
-                            //
-                        }
+            var sgfhash = checksum(sgfbuffer, 'sha256');
+
+            // upload match game to database
+            var dbres = await db.collection("match_games").updateOne(
+                { sgfhash: sgfhash },
+                {
+                    $set: {
+                        ip: req.ip, winnerhash: req.body.winnerhash, loserhash: req.body.loserhash, sgf: sgfbuffer.toString(),
+                        options_hash: req.body.options_hash,
+                        clientversion: Number(req.body.clientversion), winnercolor: req.body.winnercolor,
+                        movescount: (req.body.movescount ? Number(req.body.movescount) : null),
+                        score: req.body.score,
+                        random_seed: req.body.random_seed
                     }
-                }
+                },
+                { upsert: true }
+            );
+
+            // Not inserted, we got duplicate sgfhash, abort!
+            if (!dbres.upsertedId) {
+                return reject('Upload match with duplicate sgf');
             }
-        );
-      }
+
+            console.log(req.ip + " (" + req.headers['x-real-ip'] + ") " + " uploaded match " + sgfhash);
+            res.send("Match data " + sgfhash + " stored in database\n");
+            resolve();
+        });
     });
 
+    // prepare $inc 
+    var $inc = { game_count: 1 };
+    if (match.network1 == req.body.winnerhash)
+        $inc.network1_wins = 1;
+    else
+        $inc.network1_losses = 1;
+
+    // save to database using $inc and get modified document
+    match = (await db.collection("matches").findOneAndUpdate(
+        { _id: match._id },
+        { $inc: $inc },
+        { returnOriginal: false }  // return modified document 
+    )).value;
+
+    // get latest SPRT result
+    const sprt_result = SPRT(match.network1_wins, match.network1_losses);
+    var pending_match_index = pending_matches.findIndex(m => m._id.equals(match._id));
+
+    // match is found in pending_matches
+    if (pending_match_index >= 0) {
+        if (sprt_result === false) {
+            // remove from pending matches
+            console.log("SPRT: Early fail pop: " + JSON.stringify(pending_matches.splice(pending_match_index, 1)));
```

Comment:
A style issue: probably best to splice outside of the `console.log` to avoid someone thinking the message has no sideeffects.